File: examples/invoice-demo/frontend/src/features/invoice/actions/assignCustomer.ts
import { executeOnStream, createStream, flushStream, peekQuery } from '../../../wasm.ts';
import { assignCustomer as assignCustomerCmd } from '../../../commands/invoice/assignCustomer.ts';
import { peekInvoice } from '../reads/peekInvoice.ts';
import { isoDate } from './isoDate.ts';

interface CustomerDefaults {
  name: string;
  payment_terms_days: number;
  billing_street: string; billing_zip: string; billing_city: string; billing_country: string;
  shipping_street: string; shipping_zip: string; shipping_city: string; shipping_country: string;
}

function peekCustomer(customerId: string | null): CustomerDefaults | null {
  if (!customerId) return null;
  const rows = peekQuery(
    `SELECT customers.name, customers.payment_terms_days, ` +
    `customers.billing_street, customers.billing_zip, customers.billing_city, customers.billing_country, ` +
    `customers.shipping_street, customers.shipping_zip, customers.shipping_city, customers.shipping_country ` +
    `FROM customers WHERE customers.id = UUID '${customerId}'`,
  );
  if (rows.length === 0) return null;
  const r = rows[0];
  return {
    name: r[0] as string,
    payment_terms_days: r[1] as number,
    billing_street: r[2] as string, billing_zip: r[3] as string,
    billing_city: r[4] as string, billing_country: r[5] as string,
    shipping_street: r[6] as string, shipping_zip: r[7] as string,
    shipping_city: r[8] as string, shipping_country: r[9] as string,
  };
}

const addrIsEmpty = (inv: {
  billing_street: string; billing_zip: string; billing_city: string;
  shipping_street: string; shipping_zip: string; shipping_city: string;
}) =>
  !inv.billing_street && !inv.billing_zip && !inv.billing_city &&
  !inv.shipping_street && !inv.shipping_zip && !inv.shipping_city;

/**
 * Assigns a customer to an invoice. When the invoice's address fields are still
 * empty, pulls the customer's billing/shipping defaults into the command payload —
 * and derives a sensible date_due from their payment-terms. Existing addresses are
 * kept as-is so we never clobber user edits.
 *
 * The `AssignCustomer` intent command writes customer_id, address fields, date_due,
 * and the activity-log row atomically — callers no longer compose separate writes.
 */
export async function assignCustomer(invoiceId: string, customerId: string | null): Promise<void> {
  const inv = peekInvoice(invoiceId);
  if (!inv) return;
  const cust = peekCustomer(customerId);
  const copyAddr = cust && addrIsEmpty(inv);
  const stream = createStream(2);
  executeOnStream(stream, assignCustomerCmd({
    invoiceId,
    customerId,
    customerName: cust ? cust.name : '',
    billingStreet:   copyAddr ? cust.billing_street   : inv.billing_street,
    billingZip:      copyAddr ? cust.billing_zip      : inv.billing_zip,
    billingCity:     copyAddr ? cust.billing_city     : inv.billing_city,
    billingCountry:  copyAddr ? cust.billing_country  : inv.billing_country,
    shippingStreet:  copyAddr ? cust.shipping_street  : inv.shipping_street,
    shippingZip:     copyAddr ? cust.shipping_zip     : inv.shipping_zip,
    shippingCity:    copyAddr ? cust.shipping_city    : inv.shipping_city,
    shippingCountry: copyAddr ? cust.shipping_country : inv.shipping_country,
    dateDue: copyAddr && cust.payment_terms_days > 0
      ? isoDate(cust.payment_terms_days)
      : inv.date_due,
  }));
  await flushStream(stream);
}

File: examples/invoice-demo/commands/src/invoice/assign_customer.rs
use borsh::{BorshDeserialize, BorshSerialize};
use database::Database;
use serde::{Deserialize, Serialize};
use sql_engine::execute::Params;
use sql_engine::storage::Uuid;
use sync::command::{Command, CommandError};
use sync::zset::ZSet;
use ts_rs::TS;

use crate::helpers::{execute_sql, p_str, p_uuid, p_uuid_opt, DEMO_TENANT_ID};

/// Intent-Command: assign (or remove) a customer from an invoice.
///
/// Replaces `updateInvoiceHeader({...customer_id}) + logActivity(...)`.
/// When the invoice's address fields are still empty the client wrapper
/// copies the customer's billing/shipping defaults into the payload so the
/// server never needs an extra customer lookup. The activity-log row is
/// produced inside this command — callers no longer compose it.
///
/// `customer_name` is empty when the customer is being removed; the detail
/// string switches from "Kunde \"…\" zugewiesen" to "Kunde entfernt".
#[derive(Debug, Clone, BorshSerialize, BorshDeserialize, Serialize, Deserialize, TS)]
pub struct AssignCustomer {
    #[ts(type = "string")]
    pub id: Uuid,

    /// New customer (None = remove).
    #[ts(type = "string | null")]
    pub customer_id: Option<Uuid>,

    /// Customer display name used in the activity-log detail.
    /// Empty string when removing the customer.
    pub customer_name: String,

    // ── Address fields: final resolved values (may be copied from customer) ──
    pub billing_street: String,
    pub billing_zip: String,
    pub billing_city: String,
    pub billing_country: String,
    pub shipping_street: String,
    pub shipping_zip: String,
    pub shipping_city: String,
    pub shipping_country: String,

    /// Resolved due date (may be derived from customer payment_terms_days).
    pub date_due: String,

    /// Activity log fields.
    #[ts(type = "string")]
    pub activity_id: Uuid,
    pub timestamp: String,
}

fn detail_for(customer_name: &str) -> String {
    if customer_name.is_empty() {
        "Kunde entfernt".to_string()
    } else {
        format!("Kunde \"{customer_name}\" zugewiesen")
    }
}

impl Command for AssignCustomer {
    fn execute_optimistic(&self, db: &mut Database) -> Result<ZSet, CommandError> {
        let detail = detail_for(&self.customer_name);

        let mut acc = execute_sql(
            db,
            "UPDATE invoices SET \
             customer_id = :customer_id, \
             billing_street = :billing_street, billing_zip = :billing_zip, \
             billing_city = :billing_city, billing_country = :billing_country, \
             shipping_street = :shipping_street, shipping_zip = :shipping_zip, \
             shipping_city = :shipping_city, shipping_country = :shipping_country, \
             date_due = :date_due \
             WHERE invoices.id = :id",
            Params::from([
                p_uuid_opt("customer_id", &self.customer_id),
                p_str("billing_street", &self.billing_street),
                p_str("billing_zip", &self.billing_zip),
                p_str("billing_city", &self.billing_city),
                p_str("billing_country", &self.billing_country),
                p_str("shipping_street", &self.shipping_street),
                p_str("shipping_zip", &self.shipping_zip),
                p_str("shipping_city", &self.shipping_city),
                p_str("shipping_country", &self.shipping_country),
                p_str("date_due", &self.date_due),
                p_uuid("id", &self.id),
            ]),
        )?;

        acc.extend(execute_sql(
            db,
            "INSERT INTO activity_log (id, timestamp, entity_type, entity_id, action, actor, detail) \
             VALUES (:aid, :ts, 'invoice', :id, 'customer_assigned', 'demo', :detail)",
            Params::from([
                p_uuid("aid", &self.activity_id),
                p_str("ts", &self.timestamp),
                p_uuid("id", &self.id),
                p_str("detail", &detail),
            ]),
        )?);

        Ok(acc)
    }
}

#[cfg(feature = "server")]
mod server_impl {
    use super::*;
    use async_trait::async_trait;
    use sqlx::{MySql, Transaction};
    use sync_server_mysql::ServerCommand;

    #[async_trait]
    impl ServerCommand for AssignCustomer {
        async fn execute_server(
            &self,
            tx: &mut Transaction<'static, MySql>,
            client_zset: &ZSet,
        ) -> Result<ZSet, CommandError> {
            let detail = detail_for(&self.customer_name);

            sqlx::query(
                "UPDATE invoices SET \
                 customer_id = ?, \
                 billing_street = ?, billing_zip = ?, \
                 billing_city = ?, billing_country = ?, \
                 shipping_street = ?, shipping_zip = ?, \
                 shipping_city = ?, shipping_country = ?, \
                 date_due = ? \
                 WHERE invoices.tenant_id = ? AND invoices.id = ?",
            )
            .bind(self.customer_id.as_ref().map(|u| u.0.to_vec()))
            .bind(&self.billing_street)
            .bind(&self.billing_zip)
            .bind(&self.billing_city)
            .bind(&self.billing_country)
            .bind(&self.shipping_street)
            .bind(&self.shipping_zip)
            .bind(&self.shipping_city)
            .bind(&self.shipping_country)
            .bind(&self.date_due)
            .bind(DEMO_TENANT_ID)
            .bind(&self.id.0[..])
            .execute(&mut **tx)
            .await
            .map_err(|e| CommandError::ExecutionFailed(format!(
                "UPDATE invoice {} assign customer: {e}", self.id,
            )))?;

            sqlx::query(
                "INSERT INTO activity_log \
                 (tenant_id, id, timestamp, entity_type, entity_id, action, actor, detail) \
                 VALUES (?, ?, ?, 'invoice', ?, 'customer_assigned', 'demo', ?) \
                 ON DUPLICATE KEY UPDATE id = id",
            )
            .bind(DEMO_TENANT_ID)
            .bind(&self.activity_id.0[..])
            .bind(&self.timestamp)
            .bind(&self.id.0[..])
            .bind(&detail)
            .execute(&mut **tx)
            .await
            .map_err(|e| CommandError::ExecutionFailed(format!(
                "INSERT activity {}: {e}", self.activity_id,
            )))?;

            Ok(client_zset.clone())
        }
    }
}
